poll: add poll_notify() api and call it in all drivers.

CV-Bowen opened the pull request:
## Summary
1. add `poll_notify()` api for poll notification;
2. call `poll_notify()` in all drivers;

This patch modify some drivers' original logic:
1. Some drivers poll notify operation do not check the `fds->events` when set the `fds->revents`, that is wrong, because the app may poll success but the `fds->revents` is not the app wants. The `poll_notify()` always check the `fds->events` before set the `fds->revents`;
2. Most drivers poll notify operation do not check the count of `fds->sem`, the `poll_nofity()` always check the count of `fds->sem`, and only post it when count < 1, because there is one and only one thread is waiting for the sem (`*fds->sem` is a local variable);
3. In `poll_notify()`, always check `POLLERR `and `POLLHUP `event, and when `POLLERR `or `POLLHUP `are set, clear the `POLLIN` and `POLLOUT `event;
4. Call `poll_notify()` to do the poll notify operation, so do not need check the `fds->events` in driver when set event to eventset (eg: `eventset |= (fds->events & POLLIN)`) (this thing has been done in `poll_notify()`) , so i remove all the check code, eg: `eventset |= (fds->events & POLLIN) ==> event |= POLLIN`;   

## Impact
Only code refactor

## Testing
Pass CI

pkarashchenko commented on include/sys/poll.h:
```diff
@@ -82,6 +82,16 @@
+
+#define POLL_NOTIFY_NOCHECK (0x00000001)
+
+#define poll_notify(sfds, nfds, eventset)           \
```
```suggestion
#define poll_notify(sfds, nfds, eventset) \
```

CV-Bowen replied:
Done

pkarashchenko commented on include/sys/poll.h:
```diff
@@ -82,6 +82,16 @@
+#define poll_notify(sfds, nfds, eventset)           \
+  poll_notify_common(sfds, nfds, eventset, 0)
+
+#define poll_notify_nocheck(sfds, nfds, eventset)   \
```
```suggestion
#define poll_notify_nocheck(sfds, nfds, eventset) \
```

CV-Bowen replied:
Done

pkarashchenko commented on net/netlink/netlink_sockif.c:
```diff
@@ -595,7 +593,7 @@ static int netlink_poll(FAR struct socket *psock, FAR struct pollfd *fds,
       if (revents != 0)
         {
           fds->revents = revents;
-          nxsem_post(fds->sem);
+          poll_notify_nocheck(&fds, 1, 0);
```
Should it be
```suggestion
          poll_notify_nocheck(&fds, 1, revents);
```
?

pkarashchenko replied:
What about this place. It seems to be different from other similar places

CV-Bowen replied:
Yes, and i will delete the api poll_notify_nocheck() (only left poll_notify()), because we must check fds->events before set the fds->revents, otherwise the app may poll success but the fds->revent is not the app wants.

CV-Bowen replied:
Done

pkarashchenko commented on fs/vfs/fs_poll.c:
```diff
@@ -276,6 +277,62 @@ static inline int poll_teardown(FAR struct pollfd *fds, nfds_t nfds,
+  for (i = 0; i < nfds; i++)
+    {
+      fds = afds[i];
+      if (fds == NULL)
+        {
+          continue;
+        }
```
minor / optional:
I would better use `if (fds != NULL)` and avoid `continue`.

CV-Bowen replied:
Done

pkarashchenko reviewed (commented):
Do we have good test coverage for this change?
I see that many places did not check `fd->sem` count before posting, but now in common function it is always checked. The impact seems to be quite high from the bird eye view.

xiaoxiang781216 commented:
> Do we have good test coverage for this change? I see that many places did not check `fd->sem` count before posting, but now in common function it is always checked. The impact seems to be quite high from the bird eye view.

The check is intent: it is unnecessary to post semaphore more than once because there is one and only one thread is waiting for it:
https://github.com/apache/incubator-nuttx/blob/master/fs/vfs/fs_poll.c#L423
you can see sem is a local variable defined at line 365.

xiaoxiang781216 commented on drivers/analog/comp.c:
```diff
@@ -91,35 +91,12 @@ static const struct comp_callback_s g_comp_callback =
-
   if (eventset & POLLERR)
     {
       eventset &= ~(POLLOUT | POLLIN);
```
don't need, handled by poll_notify now. let's remove comp_pollnotify and call poll_notify directly

CV-Bowen replied:
Done

xiaoxiang781216 commented on drivers/can/can.c:
```diff
@@ -175,22 +175,7 @@ static int can_takesem(FAR sem_t *sem)
 
 static void can_pollnotify(FAR struct can_dev_s *dev, pollevent_t eventset)
```
remove?

xiaoxiang781216 commented on drivers/input/keyboard_upper.c:
```diff
@@ -105,22 +105,7 @@ static const struct file_operations g_keyboard_fops =
 
 static void keyboard_notify(FAR struct keyboard_opriv_s *opriv)
```
remove

xiaoxiang781216 commented on drivers/input/spq10kbd.c:
```diff
@@ -446,21 +446,7 @@ static int spq10kbd_interrupt(int irq, FAR void *context, FAR void *arg)
 
 static void spq10kbd_pollnotify(FAR struct spq10kbd_dev_s *priv)
```
remove

xiaoxiang781216 commented on drivers/ipcc/ipcc_poll.c:
```diff
@@ -175,32 +175,5 @@ int ipcc_poll(FAR struct file *filep, FAR struct pollfd *fds,
 
 void ipcc_pollnotify(FAR struct ipcc_driver_s *priv, pollevent_t eventset)
```
remove

xiaoxiang781216 commented on drivers/pipes/pipe_common.c:
```diff
@@ -80,35 +80,12 @@ static int pipecommon_semtake(FAR sem_t *sem)
-
   if (eventset & POLLERR)
     {
       eventset &= ~(POLLOUT | POLLIN);
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on drivers/pipes/pipe_common.c:
```diff
@@ -80,35 +80,12 @@ static int pipecommon_semtake(FAR sem_t *sem)
 static void pipecommon_pollnotify(FAR struct pipe_dev_s *dev,
```
remove

xiaoxiang781216 commented on drivers/sensors/hc_sr04.c:
```diff
@@ -275,24 +275,13 @@ static void hcsr04_notify(FAR struct hcsr04_dev_s *priv)
-          nxsem_post(fds->sem);
-        }
-    }
+  poll_notify(priv->fds, CONFIG_HCSR04_NPOLLWAITERS, POLLIN);
```
remove hcsr04_notify, call poll_notify directly

xiaoxiang781216 commented on drivers/sensors/hts221.c:
```diff
@@ -1058,24 +1058,13 @@ static void hts221_notify(FAR struct hts221_dev_s *priv)
-          nxsem_post(fds->sem);
-        }
-    }
+  poll_notify(priv->fds, CONFIG_HTS221_NPOLLWAITERS, POLLIN);
```
remove hts221_notify ditto

xiaoxiang781216 commented on drivers/sensors/lis2dh.c:
```diff
@@ -740,24 +740,13 @@ static void lis2dh_notify(FAR struct lis2dh_dev_s *priv)
-          nxsem_post(fds->sem);
-        }
-    }
+  poll_notify(priv->fds, CONFIG_LIS2DH_NPOLLWAITERS, POLLIN);
```
remove lis2dh_notify

xiaoxiang781216 commented on drivers/serial/serial.c:
```diff
@@ -171,32 +171,7 @@ static int uart_takesem(FAR sem_t *sem, bool errout)
-            }
-        }
-    }
+  poll_notify(dev->fds, CONFIG_SERIAL_NPOLLWAITERS, eventset);
```
remove uart_pollnotify

xiaoxiang781216 commented on drivers/usbdev/adb.c:
```diff
@@ -1553,22 +1553,7 @@ static void adb_char_notify_readers(FAR struct usbdev_adb_s *priv)
-            }
-        }
-    }
+  poll_notify(dev->fds, CONFIG_USBADB_NPOLLWAITERS, eventset);
```
remove adb_char_pollnotify

xiaoxiang781216 commented on drivers/usbhost/usbhost_cdcmbim.c:
```diff
@@ -405,21 +405,7 @@ static int usbhost_ctrl_cmd(FAR struct usbhost_cdcmbim_s *priv,
-            }
-        }
-    }
+  poll_notify(priv->fds, CONFIG_USBHOST_CDCMBIM_NPOLLWAITERS, POLLIN);
```
remove usbhost_pollnotify

xiaoxiang781216 commented on drivers/usbhost/usbhost_hidkbd.c:
```diff
@@ -686,21 +686,7 @@ static void usbhost_forcetake(FAR sem_t *sem)
-            }
-        }
-    }
+  poll_notify(priv->fds, CONFIG_HIDKBD_NPOLLWAITERS, POLLIN);
```
remove usbhost_pollnotify

xiaoxiang781216 commented on drivers/usbhost/usbhost_hidmouse.c:
```diff
@@ -473,21 +473,7 @@ static void usbhost_forcetake(FAR sem_t *sem)
-            }
-        }
-    }
+  poll_notify(priv->fds, CONFIG_HIDMOUSE_NPOLLWAITERS, POLLIN);
```
remove usbhost_pollnotify

xiaoxiang781216 commented on drivers/usbhost/usbhost_xboxcontroller.c:
```diff
@@ -561,8 +561,6 @@ static void usbhost_destroy(FAR void *arg)
 
 static void usbhost_pollnotify(FAR struct usbhost_state_s *priv)
 {
-  int i;
```
remove usbhost_pollnotify

xiaoxiang781216 replied:
Ok

xiaoxiang781216 commented on drivers/usbmisc/fusb301.c:
```diff
@@ -741,24 +741,13 @@ static void fusb301_notify(FAR struct fusb301_dev_s *priv)
 {
   DEBUGASSERT(priv != NULL);
 
-  int i;
```
remove fusb301_notify

xiaoxiang781216 commented on drivers/usbmisc/fusb303.c:
```diff
@@ -927,8 +927,6 @@ static int fusb303_poll(FAR struct file *filep, FAR struct pollfd *fds,
 
 static void fusb303_notify(FAR struct fusb303_dev_s *priv)
```
remove

xiaoxiang781216 commented on drivers/usrsock/usrsock_dev.c:
```diff
@@ -159,20 +159,7 @@ static bool usrsockdev_is_opened(FAR struct usrsockdev_s *dev)
 static void usrsockdev_pollnotify(FAR struct usrsockdev_s *dev,
```
remove

xiaoxiang781216 commented on fs/mqueue/mq_open.c:
```diff
@@ -373,28 +373,7 @@ static mqd_t nxmq_vopen(FAR const char *mq_name, int oflags, va_list ap)
 #if CONFIG_FS_MQUEUE_NPOLLWAITERS > 0
 void nxmq_pollnotify(FAR struct mqueue_inode_s *msgq, pollevent_t eventset)
```
remove

xiaoxiang781216 commented on fs/vfs/fs_eventfd.c:
```diff
@@ -148,22 +148,7 @@ static void eventfd_destroy(FAR struct eventfd_priv_s *dev)
 static void eventfd_pollnotify(FAR struct eventfd_priv_s *dev,
```
remove

xiaoxiang781216 commented on fs/vfs/fs_timerfd.c:
```diff
@@ -178,22 +178,7 @@ static timerfd_t timerfd_get_counter(FAR struct timerfd_priv_s *dev)
 static void timerfd_pollnotify(FAR struct timerfd_priv_s *dev,
```
remove

xiaoxiang781216 commented on net/usrsock/usrsock_poll.c:
```diff
@@ -117,8 +117,7 @@ static uint16_t poll_event(FAR struct net_driver_s *dev,
 
   if (eventset)
```
remove the check, remove line 107-114

xiaoxiang781216 commented on net/udp/udp_netpoll.c:
```diff
@@ -101,8 +101,7 @@ static uint16_t udp_poll_eventhandler(FAR struct net_driver_s *dev,
 
       if (eventset)
```
remove the check and ` & info->fds->events`

xiaoxiang781216 commented on net/usrsock/usrsock_poll.c:
```diff
@@ -265,12 +264,7 @@ static int usrsock_pollsetup(FAR struct socket *psock,
 
   /* Check if any requested events are already in effect */
 
-  if (fds->revents != 0)
```
let's create new variable(e.g. events), we shouldn't modify revents

xiaoxiang781216 commented on net/udp/udp_netpoll.c:
```diff
@@ -227,12 +226,7 @@ int udp_pollsetup(FAR struct socket *psock, FAR struct pollfd *fds)
 
   /* Check if any requested events are already in effect */
 
-  if (fds->revents != 0)
```
add events and don't change revents

xiaoxiang781216 commented on net/tcp/tcp_netpoll.c:
```diff
@@ -352,12 +351,7 @@ int tcp_pollsetup(FAR struct socket *psock, FAR struct pollfd *fds)
-
-      nxsem_post(fds->sem);
-    }
+  poll_notify(&fds, 1, 0);
```
shouldn't change revents

xiaoxiang781216 commented on net/tcp/tcp_netpoll.c:
```diff
@@ -167,8 +167,7 @@ static uint16_t tcp_poll_eventhandler(FAR struct net_driver_s *dev,
 
-          info->fds->revents |= eventset;
-          nxsem_post(info->fds->sem);
+          poll_notify(&info->fds, 1, eventset);
```
remove the check at line 162 and all `& info->fds->events`

xiaoxiang781216 commented on net/netlink/netlink_sockif.c:
```diff
@@ -594,8 +592,7 @@ static int netlink_poll(FAR struct socket *psock, FAR struct pollfd *fds,
       revents &= fds->events;
       if (revents != 0)
```
remove line 592-593

xiaoxiang781216 commented on net/local/local_netpoll.c:
```diff
@@ -125,21 +126,7 @@ void local_event_pollnotify(FAR struct local_conn_s *conn,
-            }
-        }
-    }
+  poll_notify(conn->lc_event_fds, LOCAL_NPOLLWAITERS, eventset);
```
remove local_event_pollnotify

CV-Bowen replied:
Several places call local_event_pollnotify(), seems leave it here is better.

xiaoxiang781216 commented on fs/vfs/fs_poll.c:
```diff
@@ -276,6 +277,60 @@ static inline int poll_teardown(FAR struct pollfd *fds, nfds_t nfds,
+        {
+          /* The error event must be set in fds->revents */
+
+          fds->revents |= eventset & (fds->events | POLLERR | POLLHUP);
```
remove the similar logic from all caller.

xiaoxiang781216 commented on net/local/local_netpoll.c:
```diff
@@ -215,12 +192,12 @@ int local_pollsetup(FAR struct socket *psock, FAR struct pollfd *fds)
+          shadowfds[0]        = *fds;
           shadowfds[0].fd     = 1; /* Does not matter */
-          shadowfds[0].sem    = fds->sem;
           shadowfds[0].events = fds->events & ~POLLOUT;
```
shadowfds[0].events &= ~POLLOUT;

CV-Bowen replied:
Done

xiaoxiang781216 commented on net/local/local_netpoll.c:
```diff
@@ -215,12 +192,12 @@ int local_pollsetup(FAR struct socket *psock, FAR struct pollfd *fds)
+          shadowfds[1]        = *fds;
           shadowfds[1].fd     = 0; /* Does not matter */
-          shadowfds[1].sem    = fds->sem;
           shadowfds[1].events = fds->events & ~POLLIN;
```
shadowfds[0].events &= ~POLLIN;

CV-Bowen replied:
Done

CV-Bowen commented on drivers/input/cypress_mbr3108.c:
```diff
@@ -983,21 +983,8 @@ static int mbr3108_close(FAR struct file *filep)
 
 static void mbr3108_poll_notify(FAR struct mbr3108_dev_s *priv)
```
Done

xiaoxiang781216 commented on net/rpmsg/rpmsg_sockif.c:
```diff
@@ -910,7 +890,7 @@ static int rpmsg_socket_poll(FAR struct socket *psock,
 
       if (eventset)
```
remove?

CV-Bowen replied:
Done

xiaoxiang781216 commented on net/icmpv6/icmpv6_netpoll.c:
```diff
@@ -111,8 +111,7 @@ static uint16_t icmpv6_poll_eventhandler(FAR struct net_driver_s *dev,
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on net/icmp/icmp_netpoll.c:
```diff
@@ -111,8 +111,7 @@ static uint16_t icmp_poll_eventhandler(FAR struct net_driver_s *dev,
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on net/can/can_sockif.c:
```diff
@@ -143,15 +143,14 @@ static uint16_t can_poll_eventhandler(FAR struct net_driver_s *dev,
 
       /* Awaken the caller of poll() is requested event occurred. */
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on fs/vfs/fs_eventfd.c:
```diff
@@ -540,7 +514,7 @@ static int eventfd_do_poll(FAR struct file *filep, FAR struct pollfd *fds,
 
   if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on drivers/usrsock/usrsock_dev.c:
```diff
@@ -514,7 +491,7 @@ static int usrsockdev_poll(FAR struct file *filep, FAR struct pollfd *fds,
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

CV-Bowen commented:
I will double check the code to make sure all the poll notify places have been updated, Please do not merge. @xiaoxiang781216

CV-Bowen commented:
new modified files:
        modified:   drivers/power/battery/battery_charger.c
        modified:   drivers/power/battery/battery_gauge.c
        modified:   drivers/power/battery/battery_monitor.c
        modified:   drivers/rc/lirc_dev.c
        modified:   drivers/wireless/cc1101.c
        modified:   drivers/wireless/gs2200m.c
        modified:   drivers/wireless/lpwan/sx127x/sx127x.c
        modified:   drivers/wireless/nrf24l01.c
        modified:   fs/vfs/fs_epoll.c

xiaoxiang781216 commented on drivers/serial/uart_bth4.c:
```diff
@@ -398,10 +383,10 @@ static int uart_bth4_poll(FAR struct file *filep, FAR struct pollfd *fds,
-      eventset |= (fds->events & POLLOUT);
+      eventset |= POLLOUT;
 
       if (eventset)
```
don't need

CV-Bowen replied:
Done

xiaoxiang781216 commented on drivers/sensors/sensor.c:
```diff
@@ -899,25 +883,25 @@ static int sensor_poll(FAR struct file *filep,
+          eventset |= POLLPRI;
         }
 
       if (eventset)
```
don't need

CV-Bowen replied:
Done

xiaoxiang781216 commented on drivers/net/tun.c:
```diff
@@ -1270,7 +1264,7 @@ int tun_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
+          eventset |= POLLIN;
         }
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on fs/mqueue/mq_open.c:
```diff
@@ -127,12 +127,12 @@ static int nxmq_file_poll(FAR struct file *filep,
+          eventset |= POLLIN;
         }
 
       if (eventset)
```
remove

CV-Bowen replied:
Done

xiaoxiang781216 commented on fs/vfs/fs_poll.c:
```diff
@@ -276,6 +277,60 @@ static inline int poll_teardown(FAR struct pollfd *fds, nfds_t nfds,
+
+  DEBUGASSERT(afds != NULL && nfds >= 1);
+
+  for (i = 0; i < nfds; i++)
```
```suggestion
  for (i = 0; i < nfds && eventset; i++)
```

CV-Bowen replied:
Done

CV-Bowen commented:
Delete `if (eventset)` and add this judgement to fs_poll.c
        modified:   drivers/net/tun.c
        modified:   drivers/sensors/sensor.c
        modified:   drivers/serial/uart_bth4.c
        modified:   drivers/syslog/ramlog.c
        modified:   fs/mqueue/mq_open.c
        modified:   fs/vfs/fs_poll.c
        modified:   graphics/nxterm/nxterm_kbdin.c
        modified:   net/local/local_netpoll.c

xiaoxiang781216 reviewed (approved):
LGTM

masayuki2009 commented:
@CV-Bowen @xiaoxiang781216 
I noticed that `sim:usrsocktest` failed with this PR.

```
NuttX  10.4.0 40ef5bc6db Sep 26 2022 16:59:16 sim sim
nsh> usrsocktest
Starting unit-tests...
Testing group "char_dev" =>
	Group "char_dev": [OK]
Testing group "no_daemon" =>
	Group "no_daemon": [OK]
Testing group "basic_daemon" =>
	Group "basic_daemon": [OK]
Testing group "basic_connect" =>
	Group "basic_connect": [OK]
Testing group "basic_connect_delay" =>
	Group "basic_connect_delay": [OK]
Testing group "no_block_connect" =>
	Group "no_block_connect": [OK]
Testing group "basic_send" =>
	Group "basic_send": [OK]
Testing group "no_block_send" =>
	Group "no_block_send": [OK]
Testing group "block_send" =>
	Group "block_send": [OK]
Testing group "no_block_recv" =>
	Group "no_block_recv": [OK]
Testing group "block_recv" =>
	Group "block_recv": [OK]
Testing group "remote_disconnect" =>
	[TEST ASSERT FAILED!]
		In function "remote_disconnect_poll":
		line 688: Assertion `(ssize_t)((pfd.revents & (0x01))) == (ssize_t)(((0x01)))' failed.
			got value: 0
			should be: 1
	Group "remote_disconnect": [FAILED]
Testing group "basic_setsockopt" =>
	Group "basic_setsockopt": [OK]
Testing group "basic_getsockopt" =>
	Group "basic_getsockopt": [OK]
Testing group "basic_getsockname" =>
	Group "basic_getsockname": [OK]
Testing group "wake_with_signal" =>
	Group "wake_with_signal": [OK]
Testing group "multithread" =>
	Group "multithread": [OK]
Unit-test groups done... OK:16, FAILED:1, TOTAL:17

```

CV-Bowen commented:
> @CV-Bowen @xiaoxiang781216 I noticed that `sim:usrsocktest` failed with this PR.
> 
> ```
> NuttX  10.4.0 40ef5bc6db Sep 26 2022 16:59:16 sim sim
> nsh> usrsocktest
> Starting unit-tests...
> Testing group "char_dev" =>
> 	Group "char_dev": [OK]
> Testing group "no_daemon" =>
> 	Group "no_daemon": [OK]
> Testing group "basic_daemon" =>
> 	Group "basic_daemon": [OK]
> Testing group "basic_connect" =>
> 	Group "basic_connect": [OK]
> Testing group "basic_connect_delay" =>
> 	Group "basic_connect_delay": [OK]
> Testing group "no_block_connect" =>
> 	Group "no_block_connect": [OK]
> Testing group "basic_send" =>
> 	Group "basic_send": [OK]
> Testing group "no_block_send" =>
> 	Group "no_block_send": [OK]
> Testing group "block_send" =>
> 	Group "block_send": [OK]
> Testing group "no_block_recv" =>
> 	Group "no_block_recv": [OK]
> Testing group "block_recv" =>
> 	Group "block_recv": [OK]
> Testing group "remote_disconnect" =>
> 	[TEST ASSERT FAILED!]
> 		In function "remote_disconnect_poll":
> 		line 688: Assertion `(ssize_t)((pfd.revents & (0x01))) == (ssize_t)(((0x01)))' failed.
> 			got value: 0
> 			should be: 1
> 	Group "remote_disconnect": [FAILED]
> Testing group "basic_setsockopt" =>
> 	Group "basic_setsockopt": [OK]
> Testing group "basic_getsockopt" =>
> 	Group "basic_getsockopt": [OK]
> Testing group "basic_getsockname" =>
> 	Group "basic_getsockname": [OK]
> Testing group "wake_with_signal" =>
> 	Group "wake_with_signal": [OK]
> Testing group "multithread" =>
> 	Group "multithread": [OK]
> Unit-test groups done... OK:16, FAILED:1, TOTAL:17
> ```

I will take a look, thanks.

CV-Bowen commented:
@masayuki2009 seems the poll_notify() (fs_poll()) should not clear the POLLIN if the POLLHUP set, i will discuss with @xiaoxiang781216 about the solution, thanks again.
